MsCorePkg/MacAddressEmulationDxe/MacAddressEmulationDxe.c

```diff
@@ -0,0 +1,239 @@
+  EFI_HANDLE                        *SnpHandleBuffer;
+  EFI_SIMPLE_NETWORK_PROTOCOL       *SnpInstance;
+
+  SnpInstance = NULL;
+  SnpHandleBuffer = NULL;
+
+  if (MatchFunction != NULL) {
+    Status = gBS->LocateHandleBuffer (ByProtocol, &gEfiSimpleNetworkProtocolGuid, NULL, &HandleCount, &SnpHandleBuffer);
+    ASSERT_EFI_ERROR (Status);
+    if (EFI_ERROR(Status)) {
+      DEBUG ((DEBUG_ERROR, "[%a]: Unexpected error from LocateHandleBuffer. Status=%r\n", __FUNCTION__, Status));
+    }
+  } else {
+    Status = EFI_INVALID_PARAMETER;
+  }
+
+  if (!EFI_ERROR(Status)) {
+    for (HandleIdx=0; HandleIdx < HandleCount; HandleIdx++) {
+      Status = gBS->HandleProtocol (SnpHandleBuffer[HandleIdx], &gEfiSimpleNetworkProtocolGuid, (VOID **)&SnpInstance);
+
+      if (!EFI_ERROR(Status) && MatchFunction(SnpHandleBuffer[HandleIdx], SnpInstance, MatchFunctionContext)) {
+        break;
+      } else {
+        SnpInstance = NULL;
+      }
+    }
+  }
+
+  if (SnpHandleBuffer != NULL) {
+    FreePool(SnpHandleBuffer);
+  }
+
+  return SnpInstance;
+}
+
+VOID
+SetSnpMacViaContext (
+  IN EFI_SIMPLE_NETWORK_PROTOCOL* Snp,
+  IN OUT MAC_EMULATION_SNP_NOTIFY_CONTEXT *Context
+  )
+{
+  EFI_STATUS Status = EFI_NOT_STARTED;
+  EFI_TPL OldTpl;
+
+  DEBUG ((DEBUG_ERROR, "[%a]: Start\n", __FUNCTION__));
+
+  if (Snp != NULL && Context != NULL) {
+    OldTpl = gBS->RaiseTPL(TPL_HIGH_LEVEL);
+    gBS->RestoreTPL(TPL_CALLBACK);
+
+    Status = Snp->StationAddress (Snp, FALSE, &Context->EmulationAddress);
+
+    gBS->RaiseTPL(TPL_HIGH_LEVEL);
+    gBS->RestoreTPL(OldTpl);
+  }
+
+  if (EFI_ERROR (Status)) {
+    DEBUG ((DEBUG_ERROR, "[%a]: Failed to set MAC address on SNP interface. Status=%r\n", __FUNCTION__, Status));
+  } else {
+    // Update context to indicate that we've assigned the emulation to this particular device.
+    // save the permanent address to facilitate the check above.
+    CopyMem (&Context->PermanentAddress, &Snp->Mode->PermanentAddress, NET_ETHER_ADDR_LEN);
+    Context->Assigned = TRUE;
+  }
+}
+
+/**
+  Callback that is invoked when an SNP instance is Initialized. Checks the newly installed SNP registrations (if any)
+  and updates the MAC address if a supported adapter is found.
+
+  @param[in]  Event   - The EFI_EVENT for this notify.
+  @param[in]  Context - An instance of MAC_EMULATION_SNP_NOTIFY_CONTEXT.
+
+  @retval  None
+
+**/
+VOID
+EFIAPI
+SimpleNetworkProtocolNotify (
+  IN EFI_EVENT  Event,
+  IN VOID       *Context
+  )
+{
+  MAC_EMULATION_SNP_NOTIFY_CONTEXT  *MacContext;
+  EFI_SIMPLE_NETWORK_PROTOCOL       *SnpToConfigureEmu;
+
+  DEBUG ((DEBUG_INFO, "[%a]: Start\n", __FUNCTION__));
+  
+  if (Context == NULL) {
+    DEBUG ((DEBUG_ERROR, "[%a]: Context unexpectedly null.\n", __FUNCTION__));
+    ASSERT (Context != NULL);
+    return;
+  }
+
+  MacContext = (MAC_EMULATION_SNP_NOTIFY_CONTEXT*)Context;
+
+  SnpToConfigureEmu = FindMatchingSnp(SnpSupportsMacEmuCheck, (VOID*) MacContext);
```

Comment:
I usually prefer to let the lower level functions handle these cases otherwise you end up with several layers of NULL checks which can make the code slower (depending on whether compiler can optimize them away) and harder to read.

If I add a debug warn/verbose when input null to `SetSnpMacViaContext` would that address your concern?

Reply:
But then the two functions are coupled pretty tightly no? Additionally, this is a big reason to have unit tests perform input validation, as long as we're doing unit tests that check null inputs we should be fine.